subsys/bootloader/provision/provision_flash.c

```diff
@@ -99,3 +110,144 @@ void invalidate_public_key(u32_t key_idx)
 		nrfx_nvmc_word_write((u32_t)invalidation_token, INVALID_VAL);
 	}
 }
+
+
+/** Function for reading a half-word (2 byte value) from OTP.
+ *
+ * @details The flash in OTP supports only aligned 4 byte read operations.
+ *          This function reads the encompassing 4 byte word, and returns the
+ *          requested half-word.
+ *
+ * @param[in]  ptr  Address to read.
+ *
+ * @return value
+ */
+static u16_t read_halfword(const u16_t *ptr)
+{
+	bool top_half = ((u32_t)ptr % 4); /* Addr not div by 4 */
+	u32_t target_addr = (u32_t)ptr & ~3; /* Floor address */
+	u32_t val32 = *(u32_t *)target_addr;
+
+	return (top_half ? (val32 >> 16) : val32) & 0x0000FFFF;
+}
+
+
+/** Function for writing a half-word (2 byte value) to OTP.
+ *
+ * @details The flash in OTP supports only aligned 4 byte write operations.
+ *          This function writes to the encompassing 4 byte word, masking the
+ *          other half-word with 0xFFFF so it is left untouched.
+ *
+ * @param[in]  ptr  Address to write to.
+ * @param[in]  val  Value to write into @p ptr.
+ */
+static void write_halfword(const u16_t *ptr, u16_t val)
+{
+	bool top_half = (u32_t)ptr % 4; /* Addr not div by 4 */
+	u32_t target_addr = (u32_t)ptr & ~3; /* Floor address */
+
+	u32_t val32 = (u32_t)val | 0xFFFF0000;
+	u32_t val32_shifted = ((u32_t)val << 16) | 0x0000FFFF;
+
+	nrfx_nvmc_word_write(target_addr, top_half ? val32_shifted : val32);
+}
+
+
+static const struct monotonic_counters *get_counters_struct(void)
+{
+	return (struct monotonic_counters *)
+		&p_provision_data->key_data[num_public_keys_read()];
+}
+
+
+static const struct monotonic_counter *get_counter_struct(u16_t description)
+{
+	const struct monotonic_counters *counters = get_counters_struct();
```

Comment:
The difference between `get_counters_struct` and `get_counter_struct` is confusing, could we have `get_x_counters_struct` or `get_multi_counters_struct` just to separate the names more.

Reply:
Renamed `monotonic_counters` to `counter_collection`.